intermediary.py

```diff
@@ -67,94 +69,102 @@ class Direction(enum.Enum):
-        print(f' Stop index {i}')
+    def reset_stops(self):
+        for s in self.stops:
+            s.state = StopState.Empty
 
-        self.stop_states[i].set_state(mode, intensity)
-    
+    def get_stop_by_id(self, stop_id: str):
+        stop_index = [i for i, stop in enumerate(self.stops)
+                      if stop.name == stop_id]
+        if len(stop_index):
+            return stop_index[0]
+        return None
 
-class Intermediary:
-    def __init__(self):
-        self.strip_states = {}
-        for dest, line in dest_to_strip.items():
-            if line not in self.strip_states:
-                self.strip_states[line] = RouteStripState(line[0], line[1], route_stops['RouteStrip.'+line[0].name])
+    def set_stop_state(self, stop_id: str, state: StopState,
+                       percentage: float = 0):
+        stop_index = self.get_stop_by_id(stop_id)
 
-        # mapping of 2-tuple (line, direction) to a list of pin numbers.
-        self.used_strips = {}
+        if stop_index is None:
+            return
 
+        self.stops[stop_index].state = state
+        self.stops[stop_index].percentage = percentage
+
+
+class Intermediary:
+    def __init__(self):
         self.builder = message.MessageBuilder()
-        self.serial = device.SerialDevice('COM7', 9600)
         self.data = ''
 
+        self.line_strip_mapping = {}
+
+        self.routes = {}
+        for _, (line, direction) in dest_to_strip.items():
+            if (line, direction) not in self.routes:
+                self.routes[(line, direction)] = \
+                    Route(line, direction, route_stops['RouteStrip.' + line.name])
+
     @staticmethod
     def request_train_data():
         # TODO: connect to provider server
         return data_passer.get_train_data(('SPRP', ))
 
     @staticmethod
-    def split_line_name(name):
+    def split_line_name(name: str):
         return name.replace(' Line', '').replace(' line', '').split(' - ')
 
-    def update_state(self, interval):
+    def update_state(self, interval: int):
         print('-------------')
         print('Getting data...')
         start = datetime.datetime.now()
         data = data_passer.parse_train_data(self.data, ('SPRP', ))
         print('Parsing took', datetime.datetime.now() - start)
 
-        for s in self.strip_states.values():
+        for route in self.routes.values():
             # reset all stop LED states.
-            s.clear_stops()
+            route.reset_stops()
 
         for t in data:
             t = TrainData(*t)
             print()
             print(f'Handling route {t.route_id}, trip {t.trip_id}')
             s = (self.split_line_name(train_names[t.route_id]))
-            
+
+            print(s)
+
             # get 2-tuple based on s[1], the train's destination.
-            strip = dest_to_strip[s[1]]
-            strip_state = self.strip_states[strip]
-            # print(t)
+            train_destination = dest_to_strip[s[1]]
```

Comment:
`train_destination` actually isn't the train's destination. It's the 2-tuple (Colour, Direction).